stan-dev/stan-mode
# Version 10.0.0: company, eldoc, and flycheck integration

Here is a screencast.
![example](https://user-images.githubusercontent.com/2884877/65207295-348e4c80-da5f-11e9-9f72-2eaa60f9a019.gif)

This major update adds several new packages:
- [company-stan](https://github.com/stan-dev/stan-mode/commit/1b1a537936422adb50a2b7677203ed8c37168875) for auto-completion
- [eldoc-stan](https://github.com/stan-dev/stan-mode/commit/d3c045c74cce32121e0e9c13a7222c537a14bbf7) live help for function arguments
- [flycheck-stan](https://github.com/stan-dev/stan-mode/commit/0d875ad012d85c898fe6682b62df6182295b6871) on-the-fly syntax checking with stanc

Two existing packages have been updated:
- [stan-mode](https://github.com/stan-dev/stan-mode/commit/a6104ecf33e134192a42a62c0912e72c5bc64019) for indentation and syntax highlighting
   - Deprecate # comments while handling #include differently
- [stan-snippets](https://github.com/stan-dev/stan-mode/commit/f7ccceb91039dfcff45934326d0050f7e4d1713b) for function argument completion
  - Support | as an argument separator where appropriate

All packages include automated testing (`test-*.el` files) to ensure their behaviors. See [here](https://travis-ci.org/stan-dev/stan-mode/jobs/587276519#L1509).

The minimum requirement for the Emacs version is 25.1.

The human-based testing procedures for the PR are as follows:
- Clone the development branch somewhere and install packages as described [here](https://github.com/stan-dev/stan-mode/tree/develop#via-github)
- Configure these packages in your `.emacs.d/init.el` as described [here](https://github.com/stan-dev/stan-mode/tree/develop#configuration)
- Open the `example_schools_mod_ncp.stan` example file to examine if the behaviors in the screencast reproduces:
  1. Delete the line `mu ~ normal(0,10);` in the `model` block.
  2. Type `muu ~ nor` and wait to see if `company` brings up completion candidates. Choose `normal`
  3. Hit TAB to see if `yasnippet` insert an argument template `(mu, sigma)`. The cursor should be at `mu` and allows you to enter 0. Hit TAB to move to `sigma` and enter 10. Hit TAB again to come out and enter ;.
  4. During the operations within the function arguments, the mini-buffer at the bottom should be showing argument help `(reals mu, reals sigma)` with the current argument highlighted by `eldoc`.
  5. The misspelled `muu` should be underlined by `flycheck`.
  6. Once you move the cursor to `muu`, `flycheck` should bring up "no such variable" error information.
  7. Fixing `muu` to `mu` should eliminate the underline and message.
